<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Thomas Baardemans" />


<title>RNA-seq: SARS-CoV-2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 52px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h2 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h3 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h4 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h5 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h6 {
  padding-top: 57px;
  margin-top: -57px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Thomas Baardemans</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="RNA_seq_sars_cov_2.html">RNA-seq: SARS-CoV-2</a>
</li>
<li>
  <a href="RNA_seq_method.html">RNA-seq: Method</a>
</li>
<li>
  <a href="python.html">Python</a>
</li>
<li>
  <a href="https://github.com/ThomasBaardemans">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://nl.linkedin.com/in/thomas-baardemans">
    <span class="fa fa-linkedin"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore"><strong>RNA-seq: SARS-CoV-2</strong></h1>
<h4 class="author">Thomas Baardemans</h4>
<h4 class="date">22-6-2020</h4>

</div>


<p>During my minor I attended an “Advanced Bioinformatics” course. For part of the exam each student was assigned a subset of samples from a GEO dataset to perform a RNA-seq differential expression analysis. This report uses samples from from Blanco-Melo et al., 2020 <a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a> This analysis used the DESeq2 workflow by Love et al., 2014<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a></p>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> <strong>Introduction</strong></h1>
<p>To study the effect of SARS-COV2 on the transcription of genes A549 human cell lines were infected with USA-WA1/2020 SARS-COV2 virus or mock treated. The Illumina NextSeq 500 platform was used to sequence the cDNA library. The reads were aligned to the human reference genome (hg19) using a RNA-Seq Aligment App on Basespace. The resulting table containing the readcounts was uploaded to GEO and serves as a starting point for this analysis.</p>
</div>
<div id="rna-seq-analysis" class="section level1">
<h1><span class="header-section-number">2</span> <strong>RNA-seq analysis</strong></h1>
<div id="loadinginstalling-packages" class="section level2">
<h2><span class="header-section-number">2.1</span> loading/installing packages</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#install bioconductor if it&#39;s not installed yet.</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="dt">version =</span> <span class="st">&quot;3.12&quot;</span>)

<span class="co">#install libraries if they are not installed yet</span>
<span class="kw">require</span>(<span class="st">&quot;tidyverse&quot;</span>) <span class="op">||</span><span class="st"> </span>utils<span class="op">::</span><span class="kw">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)
<span class="kw">require</span>(<span class="st">&quot;pheatmap&quot;</span>) <span class="op">||</span><span class="st"> </span>utils<span class="op">::</span><span class="kw">install.packages</span>(<span class="st">&quot;pheatmap&quot;</span>)
<span class="kw">require</span>(<span class="st">&quot;here&quot;</span>) <span class="op">||</span><span class="st"> </span>utils<span class="op">::</span><span class="kw">install.packages</span>(<span class="st">&quot;here&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;GEOquery&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;DESeq2&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;SummarizedExperiment&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#load libraries</span>
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(GEOquery)
<span class="kw">library</span>(DESeq2)
<span class="kw">library</span>(pheatmap)
<span class="kw">library</span>(here)
<span class="kw">library</span>(SummarizedExperiment)</code></pre></div>
</div>
<div id="data-downloading-and-organizing." class="section level2">
<h2><span class="header-section-number">2.2</span> Data downloading and organizing.</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Downloading expressionset and supplementary files from GEO.</span>
GSE147507 &lt;-<span class="st"> </span><span class="kw">getGEO</span>(<span class="st">&#39;GSE147507&#39;</span>, <span class="dt">destdir=</span><span class="st">&quot;.&quot;</span>, <span class="dt">GSEMatrix =</span> <span class="ot">TRUE</span>)
GSE147507_supp &lt;-<span class="st"> </span><span class="kw">getGEOSuppFiles</span>(<span class="st">&#39;GSE147507&#39;</span>, <span class="dt">makeDirectory =</span> <span class="ot">TRUE</span>, <span class="dt">baseDir =</span> <span class="kw">getwd</span>(),
  <span class="dt">fetch_files =</span> <span class="ot">TRUE</span>, <span class="dt">filter_regex =</span> <span class="ot">NULL</span>)

<span class="co">#There is human and ferret data. We select the human data. </span>
gse_human &lt;-<span class="st"> </span>GSE147507[[<span class="dv">1</span>]]</code></pre></div>
<p>Organizing and manipulating the data to make it suitable for building a sumarizedexperiment object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># First we need to obtain the countsdata,phenodata,featuredata and metadata so they can be manipulated individually later.</span>

<span class="co">#We start with assay data (countsdata)</span>
raw_counts_human &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="dt">file =</span> here<span class="op">::</span><span class="kw">here</span>(<span class="st">&quot;GSE147507&quot;</span>, <span class="st">&quot;GSE147507_RawReadCounts_Human.tsv.gz&quot;</span>))

<span class="co">#obtaining phenotypic data. contains all the info regarding each sample. </span>
phenodata_human &lt;-<span class="st"> </span><span class="kw">pData</span>(gse_human) <span class="op">%&gt;%</span><span class="st"> </span>as_tibble

<span class="co">#First column of raw_counts contains te genenames (featuredata)</span>
rowdata_human &lt;-<span class="st"> </span>raw_counts_human<span class="op">$</span>X1

<span class="co"># lastly we collect the metadata </span>
metadata &lt;-<span class="st"> </span><span class="kw">experimentData</span>(gse_human)

<span class="co">#Next we select the samples that the assignment calls for (&quot;GSE147507: GSM4462336, GSM4462337, GSM4462338, GSM4462339, GSM4462340, GSM4462341)</span>
coldata_human &lt;-<span class="st"> </span>phenodata_human[<span class="dv">21</span><span class="op">:</span><span class="dv">26</span>,]
assaydata_human &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(raw_counts_human[,<span class="dv">22</span><span class="op">:</span><span class="dv">27</span>])
<span class="kw">rownames</span>(assaydata_human) &lt;-<span class="st"> </span>raw_counts_human<span class="op">$</span>X1

<span class="co">#changing the colname to treatment to make more sense</span>
<span class="kw">colnames</span>(coldata_human)[<span class="dv">8</span>] &lt;-<span class="st"> &quot;treatment&quot;</span></code></pre></div>
</div>
<div id="generating-a-summarizedexperiment" class="section level2">
<h2><span class="header-section-number">2.3</span> Generating a summarizedexperiment</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Now we make a summarizedexperiment </span>
se_human &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> assaydata_human,
                     <span class="dt">rowData =</span> rowdata_human,
                     <span class="dt">colData =</span> coldata_human)

<span class="kw">metadata</span>(se_human)<span class="op">$</span>metadata &lt;-<span class="st"> </span>metadata <span class="co">#adding metadata manually because the summarizedexperiment won&#39;t accept it.</span>

<span class="co"># uncomment the next lines of code to inspect the assaydata, coldata and the metadata. </span>

<span class="co">#colData(se_human)</span>
<span class="co">#metadata(se_human)</span>
<span class="co">#head(assay(se_human))</span></code></pre></div>
</div>
<div id="testing-for-differential-expression-using-deseq2" class="section level2">
<h2><span class="header-section-number">2.4</span> Testing for differential expression using DESeq2</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#differential expression analysis. The only variable that is changed between samples is treatment with SARS-CoV-2 or mock treatment. </span>

<span class="co">#make DESeq2 dataset</span>
ddsSE_human &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">DESeqDataSet</span>(se_human, <span class="dt">design =</span>  <span class="op">~</span>treatment)

<span class="co">#filter all genes expressed less than 10 times.</span>
keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">counts</span>(ddsSE_human)) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">10</span>
ddsSE_human &lt;-<span class="st"> </span>ddsSE_human[keep,]

<span class="co">#perform DESeq2 analysis</span>
ddsDE_human &lt;-<span class="st"> </span><span class="kw">DESeq</span>(ddsSE_human)</code></pre></div>
</div>
<div id="exploring-the-results" class="section level2">
<h2><span class="header-section-number">2.5</span> Exploring the results</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#getting the results of the analysis</span>
res_human &lt;-<span class="st"> </span><span class="kw">results</span>(ddsDE_human)

<span class="co">#show top genes</span>
res_human[<span class="kw">order</span>(res_human<span class="op">$</span>padj), ] <span class="op">%&gt;%</span><span class="st"> </span>head</code></pre></div>
<pre><code>## log2 fold change (MLE): treatment SARS.CoV.2.infected.A549.cells vs Mock.treated.A549.cells 
## Wald test p-value: treatment SARS.CoV.2.infected.A549.cells vs Mock.treated.A549.cells 
## DataFrame with 6 rows and 6 columns
##                  baseMean    log2FoldChange              lfcSE
##                 &lt;numeric&gt;         &lt;numeric&gt;          &lt;numeric&gt;
## STC2     3472.80474970729  4.01354416339484  0.101040911737679
## LAMC2    5759.15843475204  3.70298493699246  0.111152821502871
## GPX2     8870.37949591171 -2.78195864864618 0.0859738574901862
## C15orf48 2601.41563186467  2.74951039419932 0.0894554468106483
## FASN     6880.20205343698 -2.26011225041142 0.0751376739543801
## KRT8     41319.5434574062 -2.23441250828749 0.0773697350855379
##                       stat                pvalue                  padj
##                  &lt;numeric&gt;             &lt;numeric&gt;             &lt;numeric&gt;
## STC2      39.7219709756257                     0                     0
## LAMC2     33.3143584384568 2.39231057967507e-243 1.73406632367748e-239
## GPX2     -32.3581927094959 1.06402320224979e-229 5.14171478767173e-226
## C15orf48  30.7360869821516 1.87631215639893e-207 6.80022433282883e-204
## FASN     -30.0796142795643  8.9539490879128e-199 2.59610799854944e-195
## KRT8     -28.8796711765548  2.1495181208785e-183  5.1935940330626e-180</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show distribution of p-values</span>
<span class="kw">ggplot</span>(<span class="kw">as</span>(res_human, <span class="st">&quot;data.frame&quot;</span>), <span class="kw">aes</span>(<span class="dt">x =</span> pvalue)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">binwidth =</span> <span class="fl">0.05</span>, <span class="dt">fill =</span> <span class="st">&quot;#446e9b&quot;</span>)</code></pre></div>
<div class="figure">
<img src="RNA_seq_sars_cov_2_files/figure-html/unnamed-chunk-8-1.png" alt="Histogram of p-values found in the results. Around 8000 genes have a p &lt; 0.05 indicating they are differentially expressed. Looking at the background (p &gt; 0.2) we can estimate that around 250 out of 8000 differentially expressed genes are false positives (3%)." width="672" />
<p class="caption">
Histogram of p-values found in the results. Around 8000 genes have a p &lt; 0.05 indicating they are differentially expressed. Looking at the background (p &gt; 0.2) we can estimate that around 250 out of 8000 differentially expressed genes are false positives (3%).
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#log fold change</span>
<span class="co">#get resultnames to input into LFC function</span>
<span class="kw">resultsNames</span>(ddsDE_human)</code></pre></div>
<pre><code>## [1] &quot;Intercept&quot;                                                          
## [2] &quot;treatment_SARS.CoV.2.infected.A549.cells_vs_Mock.treated.A549.cells&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#apply Log 2 fold change</span>
resLFC_human &lt;-<span class="st"> </span><span class="kw">lfcShrink</span>(ddsDE_human, <span class="dt">coef=</span><span class="st">&quot;treatment_SARS.CoV.2.infected.A549.cells_vs_Mock.treated.A549.cells&quot;</span>, <span class="dt">type=</span><span class="st">&quot;apeglm&quot;</span>)

<span class="co">#make a vulcan plot</span>
DESeq2<span class="op">::</span><span class="kw">plotMA</span>(resLFC_human, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>))</code></pre></div>
<div class="figure">
<img src="RNA_seq_sars_cov_2_files/figure-html/unnamed-chunk-9-1.png" alt="MA plot. Blue dots represent differentially expressed genes" width="672" />
<p class="caption">
MA plot. Blue dots represent differentially expressed genes
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#principal component analysis</span>
ddsDE_rlog_human &lt;-<span class="st"> </span>ddsDE_human <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rlogTransformation</span>() 


ddsDE_rlog_human <span class="op">%&gt;%</span>
<span class="st">  </span>DESeq2<span class="op">::</span><span class="kw">plotPCA</span>(<span class="dt">intgroup =</span> <span class="kw">c</span>(<span class="st">&quot;treatment&quot;</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="dt">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="dt">colour =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure">
<img src="RNA_seq_sars_cov_2_files/figure-html/unnamed-chunk-10-1.png" alt="Principal component analysis shows the biggest variance in the data is between treated and mock treated cells (as is to be expected). There is also variance between treated cells but this only accounts for 4% of the variance in the data. The mock treated cells don't show much variance between themselves." width="672" />
<p class="caption">
Principal component analysis shows the biggest variance in the data is between treated and mock treated cells (as is to be expected). There is also variance between treated cells but this only accounts for 4% of the variance in the data. The mock treated cells don’t show much variance between themselves.
</p>
</div>
</div>
<div id="heatmap" class="section level2">
<h2><span class="header-section-number">2.6</span> Heatmap</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">matrix_results &lt;-<span class="st"> </span><span class="kw">assay</span>(ddsDE_rlog_human)

ind &lt;-<span class="st"> </span>matrix_results <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rowMeans</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">order</span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>) 

top20 &lt;-<span class="st"> </span>matrix_results[ind[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>],]
heatmap &lt;-<span class="st"> </span><span class="kw">pheatmap</span>(top20,
                    <span class="dt">scale =</span> <span class="st">&quot;row&quot;</span>)
heatmap</code></pre></div>
<div class="figure">
<img src="RNA_seq_sars_cov_2_files/figure-html/unnamed-chunk-11-1.png" alt="Heatmap clustering of top 20 of up and down regulated genes." width="672" />
<p class="caption">
Heatmap clustering of top 20 of up and down regulated genes.
</p>
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1><span class="header-section-number">3</span> <strong>Conclusion</strong></h1>
<p>Alot of genes in the A549 cells get up or down regulated when infected with the virus. analyzing what pathways are associated with the differentially expressed genes belong to.</p>
</div>
<div id="bibliography" class="section level1">
<h1><span class="header-section-number">4</span> <strong>Bibliography</strong></h1>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Daniel Blanco-Melo et al., “Imbalanced Host Response to SARS-CoV-2 Drives Development of COVID-19,” Cell 181, no. 5 (May 2020): 1036-1045.e9, <a href="https://doi.org/10.1016/j.cell.2020.04.026" class="uri">https://doi.org/10.1016/j.cell.2020.04.026</a>.<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p>Love, M.I., Huber, W. &amp; Anders, S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol 15, 550 (2014). <a href="https://doi.org/10.1186/s13059-014-0550-8" class="uri">https://doi.org/10.1186/s13059-014-0550-8</a><a href="#fnref2">↩</a></p></li>
</ol>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
